File: python/bugs.py
"""Các hàm đã được sửa đúng (đã làm sạch, không còn bug)."""


def check_status(code):
    if code == 200:
        return "OK"
    return "NOT OK"


def is_valid(a, b):
    return a is not None and b is not None


def compute(value):
    return value * 2


def append_item(item, items=None):
    if items is None:
        items = []
    items.append(item)
    return items


def divide(a, b):
    if b == 0:
        return None
    return a / b


def always_true(x):
    if x not in (1, 2):
        return "luôn vào đây"
    return "không bao giờ"

File: javascript/bugs.js
// Demo các lỗi Bug (reliability) mà SonarQube phát hiện.

function checkStatus(code) {
  // Bug: dùng == thay vì === (loose equality)
  if (code == "200") {
    return "OK";
  }
  return "NOT OK";
}

function assignInCondition(x) {
  // Bug: gán trong điều kiện thay vì so sánh
  if ((x = 5)) {
    return "luôn đúng";
  }
  return "không bao giờ";
}

const config = {
  // Bug: key trùng lặp trong object literal
  timeout: 1000,
  retries: 3,
  timeout: 2000,
};

function compute(value) {
  // Bug: code không bao giờ chạy sau return (unreachable)
  return value * 2;
  console.log("đã tính xong");
}

function checkNaN(value) {
  // Bug: so sánh trực tiếp với NaN luôn false
  if (value === NaN) {
    return "là NaN";
  }
  return "không phải NaN";
}

function identicalBranches(flag) {
  // Bug: hai nhánh if/else giống hệt nhau
  if (flag) {
    return doWork(1);
  } else {
    return doWork(1);
  }
}

function doWork(n) {
  return n + 1;
}

module.exports = {
  checkStatus,
  assignInCondition,
  config,
  compute,
  checkNaN,
  identicalBranches,
};
